### ISSUE TITLE
Paginator keeps paging backwards despite `sinceId` param

### ISSUE THREAD

**Issue by janl** [neet/masto.js#775]
Confirmed against my personal timeline, running a e.g. `listHome()` Plagiator will give me all 400 timeline entries, even if I pass in a `sinceId` that is number 6 on my timeline. The Mastodon API behaves correctly and only gives me the newest 5 results, but the `Paginatior` keeps going, because the Mastodon API returns the `link` header that keeps the pagination going.

My workaround is manually breaking out of the `for await` loops:

```js
let i = 0;
let lastStatusId = Infinity
for await (const statuses of client.v1.timelines.listHome({
	limit: 5, sinceId: '109586271375908129'
})) {
	for (const status of statuses) {
		i += 1;
		console.log(status.id)
		lastStatusId = status.id
		if (status.id <= '109586271375908129') {
			break
		}
	}
	if (i % 10 == 0) {
		console.log(i)
	}
	if (lastStatusId <= '109586271375908129') {
		break
	}
}
console.log(i)
```

**Comment by eramdam** [comment]
I'm having a similar issue with the `listFollowing` endpoint. The only way to make it behave properly is to re-use the `nextParams` on the paginator and pass it into the parameters argument for `listFollowing` but given that the field is private I'm not sure if this is the intended way to do things. Here's a code snippet presenting my use-case.


```ts
 const masto = await login({
    url: "https://octodon.social",
    accessToken: "",
  });

  const account = await masto.v1.accounts.verifyCredentials();
  const followingsReference = await masto.v1.accounts.listFollowing(
    account.id,
    {
      limit: 4,
    }
  );
  const refIds = followingsReference.map((f) => f.id);
  console.log(
    "reference",
    followingsReference.map((f) => f.id)
  );
  const firstPage = masto.v1.accounts.listFollowing(account.id, {
    limit: 2,
  });
  const firstPageIds = (await firstPage).map((f) => f.id);
  console.log("first", firstPageIds);
  const secondPage = await masto.v1.accounts.listFollowing(
    account.id,
    // @ts-expect-error `nextParams` is private, this feels wrong...
    firstPage.nextParams
  );
  const secondPageIds = secondPage.map((f) => f.id);
  console.log("second", secondPageIds);
  assert.deepStrictEqual(refIds, [...firstPageIds, ...secondPageIds]);
```

**Comment by neet** [comment]
@janl Could you give me more context of what are you trying to do with `listHome`? It makes sense to me that `Paginator` created with `sinceId` goes on because if you provide it you'll get statuses that comes after it when your timeline is sorted in descending order by timestamp.

**Comment by neet** [comment]
@eramdam It looks different issue to me. You should be able to get the second page by the following code. Does it work for you?

```ts

const firstPage = masto.v1.accounts.listFollowing(me.id, { limit: 2 });
const firstPageIds = (await firstPage).map((f) => f.id);

const { value: secondPage } = await firstPage.next();
const secondPageIds = secondPage.map((f) => f.id);

assert.deepStrictEqual(refIds, [...firstPageIds, ...secondPageIds]);
```

Let me know by opening another issue if this doesn't working as you expect.

**Comment by eramdam** [comment]
> @eramdam It looks different issue to me. You should be able to get the second page by the following code. Does it work for you?
> 
> ```ts
> const firstPage = masto.v1.accounts.listFollowing(me.id, { limit: 2 });
> const firstPageIds = (await firstPage).map((f) => f.id);
> 
> const { value: secondPage } = await firstPage.next();
> const secondPageIds = secondPage.map((f) => f.id);
> 
> assert.deepStrictEqual(refIds, [...firstPageIds, ...secondPageIds]);
> ```
> 
> Let me know by opening another issue if this doesn't work as you expect.

Yep, that works! However I still feel like my issue can be valid in some use-cases, in the case of `listFollowing` the `maxId` grabbed by `nextParams` isn't very obvious so it would still be useful to be able to access those parameters "publicly". While your workaround is fine when you're running `masto` in the same place you're calling it, I can see this not working very well in a situation where `masto` is on the backend of an app and the front-end can only pass parameters 😅 And I feel like it relates to @janl's issue in that the paginator parameters might not be obvious from the get go (but that's probably Mastodon's API fault, not this package's)

**Comment by janl** [comment]
> Could you give me more context of what are you trying to do with listHome? It makes sense to me that Paginator created with sinceId goes on because if you provide it you'll get statuses that come after it in your timeline that is sorted in descending order by timestamp. Therefore the "next page" in this case should be the statuses that come before it.

My use case is this:
- when my program starts, I want to load all the posts in the timeline that I can get (normally 400).
- but to avoid unnecessary requests, I keep track on the last post id that I got from the previous run of my program
- and I’d like to be able to paginate backwards until I hit that post id but no further
- so if the last post that I have already seen (and have a copy of), is on page two, I have no need for pages three and later.

**Comment by neet** [comment]
@janl  Thank you for the explanation.

This is because `Paginator` always pages backwards regardless of whether `sinceId` is present. For now, the snippet you illustrated in the original post was one of the ways to resolve this as a workaround.

After seeing your case, I thought what you would like to do could be achieved if `Paginator` can page in a forward direction. This is one of the features that Mastodon supports while Masto.js doesn't yet. So I'm considering adding it as a new feature

I haven't yet considered the API enough, but I published some PoC to the beta channel [masto@5.2.0-rc1](https://www.npmjs.com/package/masto/v/5.2.0-rc1). Would you mind trying it out?

Here's what I prepared. Let's say we have a timeline like below:

```
+----------------+
| Home timeline  |
+----------------+
| ID: 9          |
| ID: 8          |
| ID: 7          |
| ID: 6          |
| ID: 5          |
| ID: 4          |
| ID: 3          |
| ID: 2          |
| ID: 1          |
+----------------+
```

And then fetch it by the following pseudo-code

```ts
const LAST_LOAD_ID = 5;

// backward (top to bottom)
await masto.v1.timelines
  .listHome({ limit: 2, minId: LAST_LOAD_ID })
  .toArray()
// 👎 [[4, 3], [2, 1]]

// forward (bottom to top)
await masto.v1.timelines
  .listHome({ limit: 2, minId: LAST_LOAD_ID })
  .reverse()
  .toArray()
// 👍 [[6, 7], [8, 9]]
```

The pull request is in #779 and if it works as you expect, I would like to ship it in the next stable version. I'd like to know if it solved your problem.

**Comment by janl** [comment]
Heya, I’m sorry I have not responded here. This looks like what I need.

Unfortunately, my use-case for this evaporated, so I won’t be able to put this into testing, I’m sorry!
